javascript/ql/src/semmle/javascript/dataflow/Refinements.qll

```diff
@@ -271,9 +271,27 @@ private class IndexRefinement extends RefinementCandidate, IndexExpr {
 /** A call whose arguments and/or receiver can be used in a refinement SSA node. */
 private class CallRefinement extends RefinementCandidate, CallExpr {
   CallRefinement() {
-    getAnArgument() instanceof VarAccess
-    or
-    this.(MethodCallExpr).getReceiver() instanceof VarAccess
+    exists(string name | name = getCalleeName() |
+      name = "indexOf" or
+      name = "lastIndexOf" or
+      name = "startsWith" or
+      name = "endsWith" or
+      name = "includes" or
+      name = "include" or
+      name = "contains" or
+      name = "caseInsensitiveStartsWith" or
+      name = "caseInsensitiveEndsWith" or
+      name = "caseInsensitiveContains" or
+      name = "substring" or
+      name = "substr" or
+      name = "isRelative" or
+      name = "isAbsolute"
```

Comment:
Hm, hardcoding this list in the general refinement library seems wrong.

Reply:
Yes. As you can probably tell, I've had to power through a few places to get the job done, even when it's a bit ugly.

- The list can't be derived from something like `any(StringOps::StartsWith sw).getCalleeName()` since `StringOps` is defined in terms of the DataFlow IR.
- Factoring out the list into a predicate stored somewhere else wouldn't make any real difference. It would still be hardcoded.
- Making it extendable through an abstract class doesn't matter, since `RefinementCandidate` is already abstract and easy enough to extend.
- Moving this into `TaintedPath` won't work obviously, as the entire SSA would need to be rebuilt.

In the short term, I just don't see a better way, short of accepting FPs. The list ensures that the classes in `StringOps` can be used as sanitizer guards, which seems like it could be of value to other queries, despite not being very general.

Reply:
> Factoring out the list into a predicate stored somewhere else wouldn't make any real difference. It would still be hardcoded.

True, but it could be hardcoded in a more relevant place, like `TaintTracking.qll`. Or we could define an appropriate subclass of `RefinementCandidate` there.